components/script/dom/htmlformelement.rs

```diff
@@ -417,55 +411,85 @@ impl HTMLFormElement {
 
         let mut load_data = LoadData::new(action_components, doc.get_referrer_policy(), Some(doc.url().clone()));
 
-        let parsed_data = match enctype {
-            FormEncType::UrlEncoded => {
-                load_data.headers.set(ContentType::form_url_encoded());
-
-                form_urlencoded::Serializer::new(String::new())
-                    .encoding_override(Some(self.pick_encoding()))
-                    .extend_pairs(form_data.into_iter().map(|field| (field.name.clone(), field.value_str())))
-                    .finish()
-            }
-            FormEncType::FormDataEncoded => {
-                let boundary = self.generate_boundary();
-                let mime = mime!(Multipart / FormData; Boundary =(&boundary));
-                load_data.headers.set(ContentType(mime));
-
-                self.encode_multipart_form_data(&mut form_data, None, boundary)
-            }
-            FormEncType::TextPlainEncoded => {
-                load_data.headers.set(ContentType(mime!(Text / Plain)));
-
-                self.encode_plaintext(&mut form_data)
-            }
-        };
-
         // Step 18
         let win = window_from_node(self);
         match (&*scheme, method) {
-            // https://html.spec.whatwg.org/multipage/#submit-dialog
-            (_, FormMethod::FormDialog) => return, // Unimplemented
+            (_, FormMethod::FormDialog) => {
+                // TODO: Submit dialog
+                // https://html.spec.whatwg.org/multipage/#submit-dialog
+            }
             // https://html.spec.whatwg.org/multipage/#submit-mutate-action
-            ("http", FormMethod::FormGet) | ("https", FormMethod::FormGet) => {
-                // FIXME(SimonSapin): use url.query_pairs_mut() here.
-                load_data.url.set_query(Some(&*parsed_data));
-                self.plan_to_navigate(load_data, &win);
+            ("http", FormMethod::FormGet) | ("https", FormMethod::FormGet) | ("data", FormMethod::FormGet) => {
+                load_data.headers.set(ContentType::form_url_encoded());
+                self.mutate_action_url(&mut form_data, load_data, encoding, &win);
             }
             // https://html.spec.whatwg.org/multipage/#submit-body
             ("http", FormMethod::FormPost) | ("https", FormMethod::FormPost) => {
                 load_data.method = Method::Post;
-                load_data.data = Some(parsed_data.into_bytes());
-                self.plan_to_navigate(load_data, &win);
```

Comment:
Nope, I mean the duplication of URL encoding part, i.e. `url.query_pairs_mut()....` appears in both `mutate_action_url` and `submit_entity_body`. But I think this duplication is not a problem (spec duplicates too anyway 😄)